rfd/0185-k8s-access-non-cert-routing.md

```diff
@@ -0,0 +1,247 @@
+- V(X+2).0.0:
+  - It is now "safe" to remove the old attribute based routing from the Kube
+    Forwarder. We may wish to delay this further if we have concerns about
+    older client compatability.
+
+#### Client Compatability
+
+Kubernetes API clients have been known to vary greatly in their support for 
+more unusual configuration. We have witnessed this in the past when it came to
+supporting the ability to explicitly provide the SNI to use when making
+requests.
+
+Therefore, we should ensure that Kubernetes clients and SDKs do support
+providing a base path as part of the server address (e.g they do not strip off
+the path and only use the hostname/port).
+
+The key customer we are working with has specified that they use the following
+Kubernetes clients:
+
+- `kubectl`
+- Python Kubernetes SDK
+
+Both of these clients correctly handle a base-path as configured in Kubeconfig:
+
+```yaml
+clusters:
+  - cluster:
+      server: http://127.0.0.1:8888/example/base/path
+    name: test-base-path
+```
+
+#### Analytics
+
+It would no longer be possible to track usage of Kubernetes Access for analytics
+purposes via the certificate generation event. However, we have other events
+that indicate Kubernetes Access usage (e.g an event per request), that are
+better suited for this purpose - so this seems acceptable.
+
+#### TBot UX
+
+The `tbot` configuration will be extended with a new `kubernetes/v2` service
+type that will allow you to specify multiple kubernetes clusters to generate
+within a single Kubeconfig file.
+
+To select which clusters to include within the Kubeconfig, the name or labels
+of the clusters can be specified.
+
+```yaml
+- type: kubernetes/v2
+  destination:
+    type: directory
+    path: /opt/machine-id
+  # Each selector can either be a name or a set of labels. If multiple selectors
+  # are provided, then the output will contain the union of the results of the
+  # selectors.
+  selectors:
+    - name: my-cluster
+    - labels:
+        "env": "production"
+```
+
+### Alternatives
+
+#### Leverage SNI
+
+Rather than encoding the target Kubernetes cluster into the URL, we could
+instead encode it into the SNI.
+
+This has the following limitations:
+
+- We've already witnessed limited support for specifying SNIs in Kubernetes
+  clients.
+- We already overload the SNI and ALPN with a few meanings that makes routing
+  more difficult. Adding another level to the SNI (e.g a subdomain) would not be
+  valid with the certificates we issue today.
+
+#### Custom Protocol
+
+Rather than encoding the target Kubernetes cluster into the URL, we could 
+leverage HTTP headers, or, a custom header protocol which we inject prior to the
+request.
+
+This has the following limitations:
+
+- Limited support for customisation across the various Kubernetes SDKs.
+- Injecting a custom protocol header would require the use of a local proxy,
+  which is unwieldy to manage.
+
+## Security Considerations
+
+### Per-session MFA
```

Comment:
My outstanding concern here relates to whether or not Per-Session MFA is meant to be locked to a specific target.

Today, the `KubernetesCluster` attribute seemingly has the side-effect of locking Per-Session MFA certificates to a specific target cluster. Looking back to the original Per-Session MFA RFD, the design does state that that should be the case, but says that a "TargetName" attribute should be used to enforce this - which I can't see any evidence of within the codebase today.

If locking Per-Session MFA certs to a specific cluster is intended/required, then we can retain the `KubernetesCluster` attribute, but make this intention more explicit by moving the code that enforces this into the authorizer rather than as part of the router/forwarder.